### ISSUE TITLE
Switch cameras in react

### ISSUE THREAD

**Issue by andres0397** [twilio/twilio-video.js#1336]
hello everybody, i've  seen the example by phil nash about switch cameras in vanilla javascript, but i'm trying to implement the same concept in my project but i can't, can you upload a simple sample in react ? thanks.

**Comment by andres0397** [comment]
> question

**Comment by mding5692** [comment]
You probably need to post some code here as you're still using javascript in a React.js project so it should be pretty similar

**Comment by andres0397** [comment]
hello @mding5692 here is a piece of my code,
TypeScript is not implement yet, 

this is my component participant

`
const Participant = ({ participant, isLocal, room }: Props) => {
  const [videoTracks, setVideoTracks] = useState([]) as any;
  const [audioTracks, setAudioTracks] = useState([]) as any;
  const [videoStatus, setVideoStatus] = useState("Stop Video");
  const [voiceStatus, setVoiceStatus] = useState("Unmute");
  const [devices, setDevices] = useState([]);
  const [isUsingFrontCamera, setisUsingFrontCamera] = useState(true);

  console.log(adapter.browserDetails.browser);

  const videoRef = useRef(null);
  const audioRef = useRef(null);

  const checkDevices = (devices: any) => {
    setDevices(devices.filter((d: any) => d.kind == "videoinput"));
  };

  const trackpubsToTracks = (trackMap: any) =>
    Array.from(trackMap.values())
      .map((publication: any) => publication.track)
      .filter((track) => track !== null);

  const trackSubscribed = (track: any) => {
    if (track.kind === "video") {
      setVideoTracks((videoTracks: any) => [...videoTracks, track]);
    } else if (track.kind === "audio") {
      setAudioTracks((audioTracks: any) => [...audioTracks, track]);
    }
  };

  const trackUnsubscribed = (track: any) => {
    if (track.kind === "video") {
      setVideoTracks((videoTracks: any) =>
        videoTracks.filter((v: any) => v !== track)
      );
    } else if (track.kind === "audio") {
      setAudioTracks((audioTracks: any) =>
        audioTracks.filter((a: any) => a !== track)
      );
    }
  };

  useEffect(() => {
    setVideoTracks(trackpubsToTracks(participant.videoTracks));
    setAudioTracks(trackpubsToTracks(participant.audioTracks));

    participant.on("trackSubscribed", trackSubscribed);
    participant.on("trackUnsubscribed", trackUnsubscribed);

    return () => {
      setVideoTracks([]);
      setAudioTracks([]);
      participant.removeAllListeners();
    };
  }, [participant]);

  useEffect(() => {
    const videoTrack = videoTracks[0];
    if (videoTrack) {
      videoTrack.attach(videoRef.current);
      return () => {
        videoTrack.detach();
      };
    }
  }, [videoTracks]);

  useEffect(() => {
    const audioTrack = audioTracks[0];
    if (audioTrack) {
      audioTrack.attach(audioRef.current);
      return () => {
        audioTrack.detach();
      };
    }
  }, [audioTracks]);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then(checkDevices);
  }, []);

  const handleMuteAudio = () => {
    // Voice toggle event handler
    if (voiceStatus === "Mute") {
      participant.audioTracks.forEach((audioTrack: any) => {
        audioTrack.track.disable();
      });
      setVoiceStatus("Unmute");
    } else {
      participant.audioTracks.forEach((audioTrack: any) => {
        audioTrack.track.enable();
      });
      setVoiceStatus("Mute");
    }
  };

  const handleMuteVideo = () => {
    if (videoStatus === "Stop Video") {
      participant.videoTracks.forEach((videoTrack: any) => {
        videoTrack.track.disable();
      });
      setVideoStatus("Start Video");
    } else {
      participant.videoTracks.forEach((videoTrack: any) => {
        videoTrack.track.enable();
      });
      setVideoStatus("Stop Video");
    }
  };

  const handleChangeCamera = (e: React.ChangeEvent<HTMLSelectElement>) => {
    e.persist();
    const select = e.target.value;
    console.log(select);
    if (isLocal && devices && devices.length > 1) {
      const newCamera: any = isUsingFrontCamera ? devices[1] : devices[0];
      console.log(newCamera, "this is the new camera");
      console.log(room.localParticipant.videoTracks.values(), "track values");

      setisUsingFrontCamera(!isUsingFrontCamera);

      Video.createLocalVideoTrack({
        deviceId: { exact: newCamera.deviceId },
      }).then((newVideoTrack) => {
        console.log(newVideoTrack, "eyyy new videoTrack");
        const currentVideoTracks = trackpubsToTracks(
          room.localParticipant.videoTracks
        );
        console.log(currentVideoTracks[0].track);

        //Array.from(
        //   participant.videoTracks.values()
        // ).map((trackPublication: any) => trackPublication.track);
        // console.log(currentVideoTracks, "current video tracks");
        // const newVideoTracks = [newVideoTrack].concat(
        //   videoTracks.filter((v: any, idx: number) => idx == 0)
        // );

        trackUnsubscribed(currentVideoTracks[0]);
        room.localParticipant.tracks.forEach((trackPublication: any) => {
          if (trackPublication.track.kind === "videoinput") {
            trackPublication.track.stop();
          }
        });
        room.localParticipant.unpublishTracks(currentVideoTracks);

        // trackUnsubscribed(currentVideoTracks)
        // participant.tracks.forEach((trackPublication: any) => {
        //   if (trackPublication.track.kind === "videoinput") {
        //     trackPublication.track.stop();
        //   }
        // });
        room.localParticipant.publishTrack(newVideoTrack);
        trackSubscribed(newVideoTrack);
      });
    }
  };

  return (
    <>
      <h3>{participant.identity}</h3>
      <video ref={videoRef} autoPlay={true} controls />

      {isLocal ? <button onClick={handleMuteVideo}>Mute Video</button> : null}
      {isLocal ? <button onClick={handleMuteAudio}>Mute Audio</button> : null}

      {isLocal ? (
        <select name="select" onChange={handleChangeCamera}>
          {devices &&
            devices.map((device: { deviceId?: string; label?: string }) => (
              <option key={device.label} value={device.deviceId}>
                {device.label}
              </option>
            ))}
        </select>
      ) : null}
    </>
  );
};

export default Participant;

`

**Comment by andres0397** [comment]
and jsx section 

<h3>{participant.identity}</h3>
      <video ref={videoRef} autoPlay={true} controls />

      {isLocal ? <button onClick={handleMuteVideo}>Mute Video</button> : null}
      {isLocal ? <button onClick={handleMuteAudio}>Mute Audio</button> : null}

      {isLocal ? (
        <select name="select" onChange={handleChangeCamera}>
          {devices &&
            devices.map((device: { deviceId?: string; label?: string }) => (
              <option key={device.label} value={device.deviceId}>
                {device.label}
              </option>
            ))}
        </select>
      ) : null}

**Comment by timmydoza** [comment]
Thanks for the question @andres0397!

You might be interested in our React video app: https://github.com/twilio/twilio-video-app-react

It's an example application that we build to help developers quickly get started with Twilio Video. 

It implements a camera switching feature, which might be of interest to you. See [FlipCameraButton.tsx](https://github.com/twilio/twilio-video-app-react/blob/master/src/components/MenuBar/FlipCameraButton/FlipCameraButton.tsx). It uses the [LocalVideoTrack.restart()](https://media.twiliocdn.com/sdk/js/video/releases/2.10.0/docs/LocalVideoTrack.html#restart__anchor) function to change the camera during a call. With the `LocalVideoTrack.restart()` function, there is no need to unpublish the old video track and publish a new video track.

**Comment by andres0397** [comment]
thanks @timmydoza for the support, right now i going to implement that, see you!
